### PR TITLE
support rocketMQ-3.x-plugin (from 3.0.4-open to 3.4.6.Final)

### PR DESCRIPTION
**Pull request by carlvine500**
### New feature or improvement
- Describe the details and related test reports.
add plugin from rocketMQ-3.x , because Alibaba and many other company run online tradding with it . 
I had test rocketmq-client version from 3.2.6 to 3.4.6.Final , the version below 3.2.6 is supported in theory;

8d2711a2-8f36-41bc-adbc-9799fd22666d

### CODE REVIEW

**Comment by carlvine500** [comment]
add trace image : producer & consumer
033ddc7d-0d79-4932-8d69-4942b94cc85c
45478fd8-367d-4ce9-998b-6c93e2d3b3bb

**Comment by wu-sheng** [comment]
@carlvine500 Thanks for your continuous contributions. 

@ascrutae Please review the codes first, then guide @carlvine500 to provide our auto integration test cases.

FYI， @carlvine500 here is our test report: https://github.com/SkywalkingTest/agent-integration-test-report , all the test cases are from https://github.com/SkywalkingTest/skywalking-agent-testcases

**Comment by wu-sheng** [comment]
@carlvine500 I suggest you should learn https://github.com/SkywalkingTest/skywalking-agent-testcases/tree/master/rocketMQ-scenario first, and
1. Rename `rocketMQ-scenario` to `apache-rocketMQ-scenario` or `rocketMQ-4.x-scenario`
1. Add `rocketMQ-3.x-scenario`
1. Learn from our 4.x test cases and finish your 3.x test cases.
1. @ascrutae will help you run the cases on our infra, and provide the test results and related logs in GitHub.

CC @ascrutae please provide helps.

**Comment by wu-sheng** [comment]
@carlvine500 Please follow this cn document to build the test cases: https://github.com/SkywalkingTest/skywalking-agent-testcases/blob/master/docs/how-to-write-a-plugin-testcase-cn.md  

After we have confirmed test reports, I will merge this pull request.

**Comment by wu-sheng** [comment]
@carlvine500 Do you merge the master in right way? Please be advised my pr #822 . SkyWalking is using git submodule to link protocol and UI now.

I just check you ci logs, only see the UI links.

**Comment by wu-sheng** [comment]
This is the master branch ci log: https://travis-ci.org/apache/incubator-skywalking/builds/345124146

You can find these:
```
1.30s$ git submodule update --init --recursive
Submodule 'apm-protocol/apm-network/src/main/proto' (https://github.com/apache/incubator-skywalking-data-collect-protocol.git) registered for path 'apm-protocol/apm-network/src/main/proto'
Submodule 'skywalking-ui' (https://github.com/apache/incubator-skywalking-ui) registered for path 'skywalking-ui'
Cloning into '/home/travis/build/apache/incubator-skywalking/apm-protocol/apm-network/src/main/proto'...
Cloning into '/home/travis/build/apache/incubator-skywalking/skywalking-ui'...
Submodule path 'apm-protocol/apm-network/src/main/proto': checked out '86bfe4f6635a8f71abca2383f54565b621ed3f94'
Submodule path 'skywalking-ui': checked out 'd4e49aa0581e510884e85cc145e647dcb179f5a1'
```

**Comment by wu-sheng** [comment]
@carlvine500 I found a similar issue in another committer's pc. I prefer the easiest way to delete/fork, and send a new pull request.

**Comment by wu-sheng** [comment]
The conflict triggers your ci fail. @carlvine500

**Comment by carlvine500** [comment]
@wu-sheng  sorry , I have revert to version 84358710812068cea613875ec87aba611f220b9e

**Comment by carlvine500** [comment]
add rocketMQ-3.x-plugin again

**Comment by carlvine500** [comment]
add rocketMQ-3.x-plugin again

**Comment by carlvine500** [comment]
all testCases passed , please review the code
([test result link](https://github.com/SkywalkingTest/agent-integration-test-report/blob/master/2018/2/carlvine500/testReport-master-2018-02-26-15-10.md))
98140bb3-8f84-4b52-87dc-86eafd32af7b

**Comment by wu-sheng** [comment]
@carlvine500 Can you leave your company info here: https://github.com/apache/incubator-skywalking/issues/443 . I remember you said you are using SkyWalking.

**Comment by wu-sheng** [comment]
FYI, I will submit a high priority pr today, about javadoc, which have a chance to trigger your ci failure. 

After I merge that pr, please be advised your ci status.

**Comment by carlvine500** [comment]
@wu-sheng  : Can you leave your company info here: #443 
done!

**Comment by wu-sheng** [comment]
@carlvine500 I sync master branch codes, and change the version number in your codes. Trying to fix ci.

**Comment by ascrutae** [comment]
@wu-sheng  the test case project about rocketMQ-3.x plugin has been merged. and this PR LGTM.

File: apm-sniffer/apm-sdk-plugin/rocketMQ-3.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/rocketMQ/v3/define/MQClientAPIImplInstrumentation.java [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+package org.apache.skywalking.apm.plugin.rocketMQ.v3.define;
+
+import net.bytebuddy.description.method.MethodDescription;
+import net.bytebuddy.matcher.ElementMatcher;
+import org.apache.skywalking.apm.agent.core.plugin.interceptor.ConstructorInterceptPoint;
+import org.apache.skywalking.apm.agent.core.plugin.interceptor.InstanceMethodsInterceptPoint;
+import org.apache.skywalking.apm.agent.core.plugin.interceptor.enhance.ClassInstanceMethodsEnhancePluginDefine;
+import org.apache.skywalking.apm.agent.core.plugin.match.ClassMatch;
+import org.apache.skywalking.apm.plugin.rocketMQ.v3.MessageSendInterceptor;
```
Comment (wu-sheng) [inline apm-sniffer/apm-sdk-plugin/rocketMQ-3.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/rocketMQ/v3/define/MQClientAPIImplInstrumentation.java]: This `import` should not be existed. I think your comments cause this.

File: apm-sniffer/apm-sdk-plugin/rocketMQ-3.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/rocketMQ/v3/define/MQClientAPIImplInstrumentation.java [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+import static org.apache.skywalking.apm.agent.core.plugin.bytebuddy.ArgumentTypeNameMatch.takesArgumentWithType;
+import static org.apache.skywalking.apm.agent.core.plugin.match.NameMatch.byName;
+
+/**
+ * {@link MQClientAPIImplInstrumentation} intercepts the {@link com.alibaba.rocketmq.client.impl.MQClientAPIImpl#sendMessage(String,
+ * String, com.alibaba.rocketmq.common.message.Message, com.alibaba.rocketmq.common.protocol.header.SendMessageRequestHeader,
+ * long, com.alibaba.rocketmq.client.impl.CommunicationMode, com.alibaba.rocketmq.client.producer.SendCallback,
+ * com.alibaba.rocketmq.client.impl.producer.TopicPublishInfo, com.alibaba.rocketmq.client.impl.factory.MQClientInstance,
+ * int, com.alibaba.rocketmq.client.hook.SendMessageContext, com.alibaba.rocketmq.client.impl.producer.DefaultMQProducerImpl)}
+ * method by using {@link MessageSendInterceptor}.
```
Comment (wu-sheng) [inline apm-sniffer/apm-sdk-plugin/rocketMQ-3.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/rocketMQ/v3/define/MQClientAPIImplInstrumentation.java]: Remove the link or use full name of `MessageSendInterceptor`

File: apm-sniffer/apm-sdk-plugin/rocketMQ-3.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/rocketMQ/v3/define/SendCallbackInstrumentation.java [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package org.apache.skywalking.apm.plugin.rocketMQ.v3.define;
+
+import net.bytebuddy.description.method.MethodDescription;
+import net.bytebuddy.matcher.ElementMatcher;
+import org.apache.skywalking.apm.plugin.rocketMQ.v3.OnSuccessInterceptor;
```
Comment (wu-sheng) [inline apm-sniffer/apm-sdk-plugin/rocketMQ-3.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/rocketMQ/v3/define/SendCallbackInstrumentation.java]: Same issue.

File: apm-sniffer/apm-sdk-plugin/rocketMQ-3.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/rocketMQ/v3/define/SendCallbackInstrumentation.java [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+import org.apache.skywalking.apm.agent.core.plugin.interceptor.enhance.ClassInstanceMethodsEnhancePluginDefine;
+import org.apache.skywalking.apm.agent.core.plugin.match.ClassMatch;
+
+import static net.bytebuddy.matcher.ElementMatchers.named;
+import static org.apache.skywalking.apm.agent.core.plugin.bytebuddy.ArgumentTypeNameMatch.takesArgumentWithType;
+import static org.apache.skywalking.apm.agent.core.plugin.match.HierarchyMatch.byHierarchyMatch;
+
+/**
+ * {@link SendCallbackInstrumentation} intercepts {@link com.alibaba.rocketmq.client.producer.SendCallback#onSuccess(com.alibaba.rocketmq.client.producer.SendResult sendResult)}
+ * method by using {@link OnSuccessInterceptor} and also intercepts {@link
```
Comment (wu-sheng) [inline apm-sniffer/apm-sdk-plugin/rocketMQ-3.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/rocketMQ/v3/define/SendCallbackInstrumentation.java]: Same issue.

**Comment by wu-sheng** [comment]
@carlvine500 After the final fix, I think this pr is ready to merge. Thank and congrats.

**Comment by wu-sheng** [comment]
Merged and pr done.

**Pull request merged**